src/library_openal.js

```diff
@@ -1,1622 +1,4542 @@
+    }
+
+  },
+
+  // ***************************************************************************
+  // ** ALC API 
+  // ***************************************************************************
+
+  // -------------------------------------------------------
+  // -- ALC Capture
+  // -------------------------------------------------------
+
+  // bufferSize is actually "number of sample frames", so was renamed
+  // bufferFrameCapacity here for clarity.
+  alcCaptureOpenDevice: function(pDeviceName, requestedSampleRate, format, bufferFrameCapacity) {
+
+    var resolvedDeviceName = AL.CAPTURE_DEVICE_NAME;
+
+    // NULL is a valid device name here (resolves to default);
+    if(pDeviceName !== 0) {
+      resolvedDeviceName = Pointer_stringify(pDeviceName);
+      if(resolvedDeviceName !== AL.CAPTURE_DEVICE_NAME) {
+#if OPENAL_DEBUG
+        console.error("alcCaptureOpenDevice() with invalid device name \""+resolvedDeviceName+"\"");
+#endif
+        // ALC_OUT_OF_MEMORY
+        // From the programmer's guide, ALC_OUT_OF_MEMORY's meaning is
+        // overloaded here, to mean:
+        // "The specified device is invalid, or can not capture audio."
+        // This may be misleading to API users, but well...
+        AL.alcErr = 0xA005 /* ALC_OUT_OF_MEMORY */;
+        return 0;
+      }
+    }
+
+    // Otherwise it's probably okay (though useless) for bufferFrameCapacity to be zero.
+    if (bufferFrameCapacity < 0) { // ALCsizei is signed int
+#if OPENAL_DEBUG
+      console.error("alcCaptureOpenDevice() with negative bufferSize");
+#endif
+      AL.alcErr = 0xA004 /* ALC_INVALID_VALUE */;
+      return 0;
+    }
+
+    navigator.getUserMedia = navigator.getUserMedia 
+      || navigator.webkitGetUserMedia 
+      || navigator.mozGetUserMedia 
+      || navigator.msGetUserMedia;
+    var has_getUserMedia = navigator.getUserMedia 
+      || (navigator.mediaDevices 
+      &&  navigator.mediaDevices.getUserMedia);
+
+    if (!has_getUserMedia) {
+#if OPENAL_DEBUG
+      console.error("alcCaptureOpenDevice() cannot capture audio, because your browser lacks a `getUserMedia()` implementation");
+#endif
+      // See previously mentioned rationale for ALC_OUT_OF_MEMORY
+      AL.alcErr = 0xA005 /* ALC_OUT_OF_MEMORY */;
+      return 0;
+    }
+
+    var AudioContext = window.AudioContext || window.webkitAudioContext;
+
+    if(!AL.sharedCaptureAudioCtx) {
+      try {
+        AL.sharedCaptureAudioCtx = new AudioContext();
+      } catch(e) {
+#if OPENAL_DEBUG
+        console.error("alcCaptureOpenDevice() could not create the shared capture AudioContext: " + e);
+#endif
+        // See previously mentioned rationale for ALC_OUT_OF_MEMORY
+        AL.alcErr = 0xA005 /* ALC_OUT_OF_MEMORY */;
+        return 0;
+      }
+    }
+
+    var outputChannelCount;
+
+    switch (format) {
+    case 0x10010: /* AL_FORMAT_MONO_FLOAT32 */
+    case 0x1101:  /* AL_FORMAT_MONO16 */
+    case 0x1100:  /* AL_FORMAT_MONO8 */
+      outputChannelCount = 1;
+
```

Comment:
This kind of pattern is explicitly reverting this earlier PR: https://github.com/kripken/emscripten/pull/4275/files#diff-36fcd036fa88446c972f56ca850c357fL1387. I see your documentation did have a mention about being sensitive to garbage collection pressure, so it would be good to retain the garbage free operation from PR #4275?